src/gui_common/CustomDialog.cs

```diff
@@ -0,0 +1,489 @@
+                new Vector2(adjustedRect.Position.x - scaleborderSize, adjustedRect.Position.y - scaleborderSize),
+                new Vector2(adjustedRect.Size.x + scaleborderSize * 2, adjustedRect.Size.y + scaleborderSize * 2));
+        }
+
+        return adjustedRect.HasPoint(point);
+    }
+
+    /// <summary>
+    ///   Overrides the minimum size to account for default elements (e.g title, close button, margin) rect size
+    ///   and for the other custom added contents on the window.
+    /// </summary>
+    public override Vector2 _GetMinimumSize()
+    {
+        var margin = GetConstant("custom_margin", "Dialogs");
+        var font = GetFont("custom_title_font", "WindowDialog");
+
+        var buttonWidth = closeButton?.GetCombinedMinimumSize().x;
+        var titleWidth = font.GetStringSize(TranslationServer.Translate(windowTitle)).x;
+        var buttonArea = buttonWidth + (buttonWidth / 2);
+
+        var contentSize = Vector2.Zero;
+
+        for (int i = 0; i < GetChildCount(); ++i)
+        {
+            var child = GetChildOrNull<Control>(i);
+
+            if (child == null || child == closeButton || child.IsSetAsToplevel())
+                continue;
+
+            var childMinSize = child.GetCombinedMinimumSize();
+
+            contentSize = new Vector2(
+                Mathf.Max(childMinSize.x, contentSize.x),
+                Mathf.Max(childMinSize.y, contentSize.y));
+        }
+
+        // Redecide whether the largest rect is the default elements' or the contents'
+        return new Vector2(Mathf.Max(2 * buttonArea.GetValueOrDefault() + titleWidth,
+            contentSize.x + margin * 2), contentSize.y + margin * 2);
+    }
+
+    /// <summary>
+    ///   Called after the popup is made visible.
+    /// </summary>
+    protected virtual void OnShown()
+    {
+        // TODO: implement default show animation(?)
+    }
+
+    /// <summary>
+    ///   Called after popup is made invisible.
+    /// </summary>
+    protected virtual void OnHidden()
+    {
+        // TODO: add proper close animation
+        closeHovered = false;
+    }
+
+    /// <summary>
+    ///   Evaluates what kind of drag type is being done on the window based on the current mouse position.
+    /// </summary>
+    private DragType DragHitTest(Vector2 position)
+    {
+        var result = DragType.None;
+
+        if (Resizable)
+        {
+            var scaleborderSize = GetConstant("custom_scaleborder_size", "WindowDialog");
+            var titleBarHeight = GetConstant("custom_titlebar_height", "WindowDialog");
+
+            if (position.y < (-titleBarHeight + scaleborderSize))
+            {
+                result = DragType.ResizeTop;
+            }
+            else if (position.y >= (RectSize.y - scaleborderSize))
+            {
+                result = DragType.ResizeBottom;
+            }
+
+            if (position.x < scaleborderSize)
+            {
+                result |= DragType.ResizeLeft;
```

Comment:
Is this code derived from Godot code or our own reimplementation?
I'm slightly getting the former vibe with how much complex logic there is in this class...
If that's the case then we need to more explicitly put the copyright notice in this file from Godot and say that this is a modified version.

Reply:
Yes, though a heavily modified one. Will add it.